app/src/sharedTest/java/org/oppia/app/utility/DragViewAction.kt

```diff
@@ -0,0 +1,173 @@
+package org.oppia.app.utility
+
+import android.os.SystemClock
+import android.view.View
+import android.view.ViewConfiguration
+import androidx.recyclerview.widget.RecyclerView
+import androidx.test.espresso.PerformException
+import androidx.test.espresso.UiController
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.action.CoordinatesProvider
+import androidx.test.espresso.action.MotionEvents
+import androidx.test.espresso.action.PrecisionDescriber
+import androidx.test.espresso.action.Swiper
+import androidx.test.espresso.matcher.ViewMatchers.isCompletelyDisplayed
+import androidx.test.espresso.util.HumanReadables
+import org.hamcrest.Matcher
+import org.hamcrest.core.AllOf.allOf
+
+/**
+ * The number of move events to send for each drag.
+ */
+private const val DRAG_STEP_COUNT = 10
+
+/**
+ * Length of time a drag should last for, in milliseconds.
+ */
+private const val DRAG_DURATION = 1500
+
+/**
+ * Duration between the last move event and the up event, in milliseconds.
+ */
+private const val WAIT_BEFORE_SENDING_UP = 400
+
+/**
+ * A custom [ViewAction] that can be replicate the long press & drag action for espresso.
+ *
+ * Reference: https://android.googlesource.com/platform/frameworks/base/+/2ff41d4/core/tests/coretests/src/android/widget/espresso/DragAction.java
+ */
+class DragViewAction(
+  private val startCoordinatesProvider: CoordinatesProvider,
+  private val endCoordinatesProvider: CoordinatesProvider,
+  private val precisionDescriber: PrecisionDescriber
+) : ViewAction {
+
+  override fun getDescription(): String = "long press and drag"
+
+  override fun getConstraints(): Matcher<View> = allOf(isCompletelyDisplayed())
+
+  override fun perform(uiController: UiController, view: View) {
+    val startCoordinates = startCoordinatesProvider.calculateCoordinates(view)
+    val endCoordinates = endCoordinatesProvider.calculateCoordinates(view)
+    val precision = precisionDescriber.describePrecision()
+    val downEvent = MotionEvents.sendDown(uiController, startCoordinates, precision).down
+    var status: Swiper.Status
+
+    try {
+      // Factor 1.5 is needed, otherwise a long press is not safely detected.
+      val longPressTimeout = (ViewConfiguration.getLongPressTimeout() * 1.5f).toLong()
+      uiController.loopMainThreadForAtLeast(longPressTimeout)
+
+      val steps = interpolate(startCoordinates, endCoordinates)
+      val delayBetweenMovements = DRAG_DURATION / steps.size
+
+      steps.forEachIndexed { index, step ->
+        if (!MotionEvents.sendMovement(uiController, downEvent, step)) {
+          MotionEvents.sendCancel(uiController, downEvent)
+          status = Swiper.Status.FAILURE
+        }
+        val desiredTime: Long = downEvent.downTime + delayBetweenMovements * index
+        val timeUntilDesired = desiredTime - SystemClock.uptimeMillis()
+        if (timeUntilDesired > 10) {
+          // If the wait time until the next event isn't long enough, skip the wait
```

Comment:
A bunch of this code still seems copied from DragAction. Was it rewritten?

Reply:
It was rewritten but all the things rhat i did here has to be done otherwise two different events gets overlapped

Reply:
All of the code is quite basic and it's to be done i can add a gif of that if i don't use this code